File: src/lib/identities.ts
import { invoke } from "@tauri-apps/api/core";
import { isTauri } from "@tauri-apps/api/core";

// Thin wrappers around src-tauri/src/identities.rs.
// Mutes are account-scoped and keyed by the root message's Message-ID (angle
// brackets stripped, the same format stored in the cache).

export function muteThread(accountId: string, messageId: string): Promise<void> {
  return invoke("mute_thread", { accountId, messageId });
}

export function unmuteThread(accountId: string, messageId: string): Promise<void> {
  return invoke("unmute_thread", { accountId, messageId });
}

export function listMutedThreads(accountId: string): Promise<string[]> {
  return invoke("list_muted_threads", { accountId });
}

// Send-as identities (aliases) for an account. The `address` is the From
// address shown in outgoing mail; authentication still uses the account's
// stored credential.

export interface IdentitySummary {
  account_id: string;
  address: string;
  display_name: string | null;
  signature: string | null;
}

export function listIdentities(accountId: string): Promise<IdentitySummary[]> {
  if (!isTauri()) return Promise.resolve([]);
  return invoke("list_identities", { accountId });
}

export function addIdentity(
  accountId: string,
  address: string,
  displayName: string | null,
  signature: string | null,
): Promise<void> {
  return invoke("add_identity", { accountId, address, displayName, signature });
}

export function deleteIdentity(accountId: string, address: string): Promise<void> {
  return invoke("delete_identity", { accountId, address });
}

File: src-tauri/src/identities.rs
use serde::Serialize;
use crate::cache::{self, IdentityRecord};

#[derive(Debug, Serialize)]
pub struct IdentitySummary {
    pub account_id: String,
    pub address: String,
    pub display_name: Option<String>,
    pub signature: Option<String>,
}

/// Adds or updates a send-as identity for an account. The `address` is the
/// From address that will appear in the message; authentication still uses
/// the account's stored credential. Idempotent: re-adding the same address
/// updates `display_name` and `signature` in place.
#[tauri::command]
pub fn add_identity(
    account_id: String,
    address: String,
    display_name: Option<String>,
    signature: Option<String>,
) -> Result<(), String> {
    let conn = cache::open()?;
    cache::upsert_identity(
        &conn,
        &IdentityRecord {
            account_id,
            address,
            display_name,
            signature,
            created_at: chrono::Utc::now().to_rfc3339(),
        },
    )
}

/// Returns all send-as identities for the given account.
#[tauri::command]
pub fn list_identities(account_id: String) -> Result<Vec<IdentitySummary>, String> {
    let conn = cache::open()?;
    cache::list_identities(&conn, &account_id).map(|records| {
        records
            .into_iter()
            .map(|r| IdentitySummary {
                account_id: r.account_id,
                address: r.address,
                display_name: r.display_name,
                signature: r.signature,
            })
            .collect()
    })
}

/// Removes a send-as identity. Silently succeeds if the address isn't found.
#[tauri::command]
pub fn delete_identity(account_id: String, address: String) -> Result<(), String> {
    let conn = cache::open()?;
    cache::delete_identity(&conn, &account_id, &address)
}

/// Mutes a thread so that new arrivals in it are silently marked seen without
/// triggering a notification. `message_id` is the root message's Message-ID
/// header (angle brackets stripped, the same format stored in cache). The mute
/// is account-scoped so muting on one account doesn't affect another.
#[tauri::command]
pub fn mute_thread(account_id: String, message_id: String) -> Result<(), String> {
    let conn = cache::open()?;
    cache::mute_thread_in(&conn, &account_id, &message_id)
}

/// Removes the mute from a thread. Silently succeeds if the thread wasn't muted.
#[tauri::command]
pub fn unmute_thread(account_id: String, message_id: String) -> Result<(), String> {
    let conn = cache::open()?;
    cache::unmute_thread_in(&conn, &account_id, &message_id)
}

/// Returns `true` if the thread rooted at `message_id` is currently muted
/// for this account.
#[tauri::command]
pub fn is_thread_muted(account_id: String, message_id: String) -> Result<bool, String> {
    let conn = cache::open()?;
    cache::is_thread_muted(&conn, &account_id, &message_id)
}

/// Returns all muted thread root Message-IDs for an account.
#[tauri::command]
pub fn list_muted_threads(account_id: String) -> Result<Vec<String>, String> {
    let conn = cache::open()?;
    cache::list_muted_threads(&conn, &account_id)
}
